src/main/java/com/atlassian/db/replica/api/AuroraLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+package com.atlassian.db.replica.api;
+
+import com.atlassian.db.replica.internal.MonotonicMemoryCache;
+import com.atlassian.db.replica.spi.Cache;
+import com.atlassian.db.replica.spi.ReplicaConsistency;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public class AuroraLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraLsnReplicaConsistency build() {
+            return new AuroraLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
```

Comment:
I've checked the app-side, and we log the exception on the cache level before we re-throw it to the Replica consistency, so we are not losing the exception. I just wonder why to catch and handle it here. We don't do it in PessimisticPropagationConsistency. 🤔

Reply:
It's because `clock.instant()` is much less error-prone than `queryMainDbLsn(main)` :)

Reply:
Yes. But here, we wrap `put` as well.

Reply:
In the case of put failure we should also invalidate cache. Why it shouldn't be part of consistency implementation?

Reply:
I don't say it shouldn't be here. I just noticed we didn't do it in the previous implementation.

Reply:
IMO put failure may mean it timeouts, and we don't know if it propagated. Then we can try to reset it. It may fail or not. I'm not sure what value we get from it.

Anyway, we should know the value we can document if with a test.